### ISSUE TITLE
Scrapyd poller stops getting new jobs under load

### ISSUE THREAD

**Issue by hbbtstar** [scrapy/scrapyd#225]
I've been scratching my head for a bit with a problem and I was wondering if I could solicit some advice.

So we use scrapyd to pop jobs off a queue that implements ISpiderQueue, using the implementation of QueuePoller found here: 

https://github.com/scrapy/scrapyd/blob/master/scrapyd/poller.py

for the most part, this works just fine. When there are lots of jobs, however, the poller appears to stop taking jobs off the queue, causing the queue size to skyrocket. I originally thought this might be due to max_proc / max_proc_per_cpu, but even once the number of jobs has fallen under that threshold the poller doesn't grab new jobs; once the poller stops taking jobs, it won't take any more until scrapy is restarted via supervisor, even if the number of jobs falls under the max_proc limit. 

I've set log level to DEBUG, but nothing jumps out at me: I don't see any exceptions, warnings, or errors. this is the final line of the log when this happens:

``` 2017-04-06T09:39:26+0000 [-] Process finished:  project='default' spider='generic_spider' job='80222d1c1aac11e783660a334fc553ef' pid=7140 log='/var/log/scrapy/default/generic_spider/80222d1c1aac11e783660a334fc553ef.log' items=None ```

I don't think it's a load issue: once scrapy is restarted, the machine catches up on the queue backlog easily, and the CPU / memory on the machine isn't maxed out. 

Here's our implementation of the Poller, it's very similar to the one in the scrapyd repo but maybe I've done something wrong:

https://gist.github.com/hbbtstar/7f874fc08371103ae42b81e609906bea

I'd appreciate any pointers as to where to investigate next, as this leaves me pretty stumped. Thanks!

**Comment by Digenis** [comment]
Hi @hbbtstar,

It's nice to see people experimenting with scrapyd components.
Issues you open may offer useful feedback and improve scrapyd.

Now in regards to your issue.
Do other scrapyd components work?
I'm asking because you may be enqueuing jobs and checking queue size directly in redis.
Does the scrapyd website and list jobs webservice work?
Consider adding some logging in the poller.
Can you reproduce the issue with the builtin poller?

**Comment by hbbtstar** [comment]
Hi Digenis!

Thanks for replying: I closed the ticket because I ended up figuring it out, and it wasn't related to the poller, so I didn't think it was fair to keep a bug open and make you guys investigate it. What happened was that there was actually an error scenario in our redis queue implementation, so when it threw an exception on .pop and .add it would take out the poller and the service would have to be restarted.

Thank you for replying, and sorry for dragging you down a rabbit hole!
